[Zheng Jiarui] iP

carriezhengjr opened the pull request:
## CaCa 🤖
> "It is our attitude at the beginning of a difficult task which, more than anything else, will affect its successful outcome." – William James ([source](https://www.brainyquote.com/quotes/william_james_157168))

Start your task easily and boost your productivity with CaCa 🤖, a chatbot which frees your mind 🤯 of having to remember  and keep track of overwhelming things you need to do! ✅

CaCa is
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/carriezhengjr/ip/releases/tag/A-Jar).
2. open a command window in that same folder.
3. run the command `java -jar {filename}.jar`. 
    - e.g., `java -jar caca.jar` (i.e., run the command in the same folder as the jar file)
4. let it manage your tasks for you! 😉

And it is **FREE**!

Features:
- [x] Managing tasks (add, delete, mark, unmark, list, find)
- [x] Managing deadlines
- [x] Managing events

Look out for more features in future! 😎

<hr>

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

carriezhengjr committed:
Level 1: customised chatbot name, display; added greeting, echo, exit

carriezhengjr committed:
Level 2: added ability to store and display items

carriezhengjr committed:
Level 3: added ability to mark tasks as done or undone

carriezhengjr committed:
Level 4: added support for tracking ToDos, Events, Deadlines

carriezhengjr committed:
Ignore .class files

carriezhengjr committed:
Inform user when task list is empty

carriezhengjr committed:
Resolved formatting issue; deleted logo to pass testing

carriezhengjr committed:
A-TextUiTesting using I/O redirection technique

carriezhengjr committed:
Level 5: Handle Errors

carriezhengjr committed:
Level 6: added support for deleting tasks

carriezhengjr committed:
Added back CaCa logo, able to pass test with "diff -w" instead of "diff

carriezhengjr committed:
Added reference for CaCa logo as a comment

carriezhengjr committed:
Created individual functions for greeting, bye, listTasks, markTask and unmarkTask to tidy up code in main

carriezhengjr committed:
Edit JavaDoc

Edit JavaDoc for isValid method.

carriezhengjr committed:
Update README.md

Add CaCa introduction and features

carriezhengjr committed:
Add author and version for all classes

Edit JavaDoc of all classes according to requirements by adding author and version

carriezhengjr committed:
Add description for CaCa class

Add features with detailed description

carriezhengjr committed:
Improve code quality

Change constants to caps

carriezhengjr committed:
Improve code readability

Improve code quality by declaring variables and use variable name when possible.

carriezhengjr committed:
Improve code readability

Improve code quality by changing todo, deadline and event into methods to be used in main.

carriezhengjr committed:
Declare variables as final, add JavaDoc

Declare LINE and array tasks as final. Add previously missing JavaDoc for @throws. Renamed some variable for a better name.

carriezhengjr committed:
Change delete into a method

Improve code quality by converting delete into a method to be used in main.

carriezhengjr committed:
Improve readability with line break

carriezhengjr committed:
Add finally statement for try-catch block

carriezhengjr committed:
Edit style to follow coding standard

carriezhengjr committed:
Use default branch in statements

Add else statement for if-else block; add JavaDoc for isValid; improve code readability

carriezhengjr committed:
Improve code structure and readability

carriezhengjr committed:
Level 7: save tasks in hard disk

Tasks entered by user is not stored in memory.

Storing task inputs could allow users to review and return to
the previous task list created more conveniently without the need
to add all the tasks again.

Let's,
* create a Storage class to save the tasks in hard disk
automatically whenever the task list changes
* load data whenever CaCa program runs.

carriezhengjr committed:
Level 8: create date and time in correct format

Dates and times for deadline and event are treated as strings by CaCa.

CaCa should understand dates and times in the correct format
so that date/time inputs are more meaningful.

Let's,
* teach CaCa how to understand dates and times by storing them
as LocalDateTime
* allow date and time inputs in the format dd/MM/yyyy HHmm
* print date and time in the format MMM dd yyyy HH:mm

carriezhengjr committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Level 8: create date and time in correct format

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

carriezhengjr committed:
Add test cases for Level-8

The test files input.txt and EXPECTED.TXT do not contain
test inputs to test functions implemented in Level 8 (Dates and Times).

Tests should be included to test the behaviour of program
for Level-8 and also in future modifications.

Let's update input.txt and EXPECTED.TXT to include those tests.

carriezhengjr committed:
A-MoreOOP: Use More OOP

Some classes and main program contain long blocks of code and logics.

Code could be refactored to extract out closely related code
as classes.

Let's extract the classes below
* Ui.java, which handles interactions with user
* Storage.java, which handles loading and saving tasks in local file
* Parser.java, which handles user command with logics
* TaskList.java, which contains operations to modify task list

carriezhengjr committed:
A-Packages: organise classes into packages

All classes are not divided into packages.

Classes should be organised into packages for easier management.

Let's divide classes into multiple packages.

carriezhengjr committed:
Merge remote-tracking branch 'upstream/add-gradle-support'
Add Gradle support from https://github.com/nus-cs2103-AY2223S1/ip

carriezhengjr committed:
A-JUnit: Add JUnit tests

There are no JUnit tests for automated testing.

JUnit is an important unit testing framework for testing projects
to ensure each class and method behaves correctly.

Let's add JUnit tests for some classes as automated unit testing.

carriezhengjr committed:
A-JavaDoc: add JavaDoc comments

Some JavaDoc comments are missing,
some are not in the correct format.

There is a need to provide clear documentation of the classes
and methods for both developer-as-user and developer-as-maintainer.

Let's add more JavaDoc comments and improve on existing ones.

carriezhengjr committed:
A-CodingStandard: follow the coding standard

Some blocks of code do not follow
the required coding standard closely.

Code should be neat and follows a consistent style
for better readability and code quality.

Let's tweak the code to comply more closely
with the given coding standard.

carriezhengjr committed:
Level-9: allow users to find a task with a keyword

Tasks added into the list are hard to find manually,
especially if the list is long.

Allowing users to enter keywords to find a task will
improve CaCa functionalities and convenience.

Let's add a find operation that CaCa can perform.

carriezhengjr committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  A-JavaDoc: add JavaDoc comments

carriezhengjr committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  A-CodingStandard: follow the coding standard

carriezhengjr committed:
Add tests for findTask

There are no tests to verify the behaviour of findTask method.

This could lead to potential bugs not discovered.

Let's add inputs and expected outputs for findTask.

carriezhengjr committed:
A-CheckStyle: use CheckStyle

One can only check for style errors manually.

Adding CheckStyle which is a static analysis tool
can help to check Java code against style rules more easily.

Let's use CheckStyle plugin with Gradle
to check for style errors more conveniently.

yellow-294 commented on src/main/java/caca/Storage.java:
```diff
@@ -0,0 +1,189 @@
+
+        if (file.exists()) {
+
+            String openFileMsg = "Opening the file that stores all your tasks...\n\n";
+            String showTaskmsg = "These are the tasks in your list:";
```
Could it be possible to let ur Ui class handle all your display messages?

carriezhengjr replied:
I'll try to see how I can improve on that. Thank you :)

yellow-294 commented on src/main/java/caca/Parser.java:
```diff
@@ -0,0 +1,129 @@
+                TaskList.hasDescription(command);
+
+                TaskList.addDeadline(command[1]);
+
+                storage.updateFile(tasks);
+
+            } else if (commandType.equals("event")) {
+                // Checks for valid description, i.e. not empty or blank, before adding event.
+                TaskList.hasDescription(command);
+
+                TaskList.addEvent(command[1]);
+
+                storage.updateFile(tasks);
+
+            } else if (commandType.equals("list")) {
+                TaskList.listTasks();
+
+            } else if (commandType.equals("mark")) {
+                TaskList.markTask(command[1]);
+
+                storage.updateFile(tasks);
+
+            } else if (commandType.equals("unmark")) {
+                TaskList.unmarkTask(command[1]);
+
+                storage.updateFile(tasks);
+
+            } else if (commandType.equals("delete")) {
+                TaskList.deleteTask(command[1]);
+
+                storage.updateFile(tasks);
+
+            } else if (commandType.equals("find")) {
+                TaskList.findTask(command[1]);
+
+            } else {
+                // Invalid input.
+                System.out.println("OOPS!!! I'm sorry, but I don't know what that means :-(");
+
+            }
```
The code here is very neat and easy to read. Good Job!

yellow-294 commented on src/main/java/caca/Storage.java:
```diff
@@ -0,0 +1,189 @@
+                String description = taskDetails[2];
+
+                // Stores all the tasks from file into the array.
+                switch (taskType) {
+                case "T":
+                    tasks.addTask(new Todo(description, isDone));
+                    break;
+                case "D":
+                    String by = taskDetails[3];
+                    tasks.addTask(new Deadline(description, by, isDone));
+                    break;
+                case "E":
+                    String at = taskDetails[3];
+                    tasks.addTask(new Event(description, at, isDone));
+                    break;
+                default:
+                    break;
+                }
+
+            }
```
Elegant use of Switch. Good Job!

yellow-294 commented on src/main/java/caca/TaskList.java:
```diff
@@ -0,0 +1,285 @@
+        Task taskToAdd = new Todo(taskDescription);
+        tasks.add(taskToAdd);
+
+        System.out.println("Got it. I've added this task:");
+        System.out.println(taskToAdd);
+        System.out.printf("Now you have %d tasks in the list.\n", tasks.size());
```
Similarly to my above comment, do you feel you could make use of your Ui class to handle all display messages?

carriezhengjr replied:
Will try to improve on that, thanks! 😄

yellow-294 reviewed (commented):
Overall code is neat.
I believe you could still use your Ui class better.
Otherwise good job!

carriezhengjr committed:
Level 10: add a GUI to project

Project displays all output and interactions with user in terminal.

This is not aesthetically pleasing and could be improved with GUI.

Let's add a GUI to CaCa and use JavaFX technology to implement it.

carriezhengjr committed:
Update README.md

README.md is not up-to-date with the most recent functions.

Updating README allows users to understand the project better 
and know how to download and run the program.

Let's ,
* update the user guide
* add in more information to promote CaCa.

carriezhengjr committed:
Update README.md

README seems wordy 
especially ending off with long user guide.

Adding some images and emojis could attract 
more attention and interest of potential users.

Let's add CaCa and user images 
to introduce the application better.

carriezhengjr committed:
A-Assertions: Add assertions

Some assumptions are made in the code but not checked.

Java assertions can verify assumptions
about the program state at runtime.

Let's use assertions to the code.

carriezhengjr committed:
A-CodeQuality: improve code quality

Some methods do not follow OOP closely and may be hard to follow.

Refactoring could improve code structure and quality.

Let's refactor to improve the code quality
(mainly in Parser class and TaskList class) by
* consolidating duplicate conditional fragments
* extracting methods
* adding any corresponding tests for new behaviours or methods

carriezhengjr committed:
Merge pull request #1 from carriezhengjr/branch-A-CodeQuality

A-CodeQuality: improve code quality

carriezhengjr committed:
Merge branch 'master' into branch-A-Assertions

* master:
  A-CodeQuality: improve code quality

# Conflicts:
#	src/main/java/caca/TaskList.java

carriezhengjr committed:
Merge pull request #2 from carriezhengjr/branch-A-Assertions

Merge A-Assertions to master

carriezhengjr committed:
A-CI: set up CI

There is no Continuous integration in the project,
so we have to build the project manually each time.

With CI, integration, building, and testing can
happen automatically after each code change.

Let's use GitHub Actions to set up CI.

carriezhengjr committed:
Add reference for CI

CI file code is copied from example given,
but does not contain any citation.

We need to give credit for reused or referenced code.

Let's add a comment for the reused code block.

carriezhengjr committed:
C-DetectDuplicates: deal with duplicate tasks

The same task can be added multiple times to the task list.

The task list may contain many duplicate tasks
and will be hard to manage, i.e. mark, unmark and delete.

Let's
* edit the addTask method to detect duplicate tasks using stream.
* add tests to test simple duplicate tasks with no task type required.
More tests with task types, i.e. todo, deadline and event
will be added in later commits.

carriezhengjr committed:
Add tests to check duplicate method

There are no unit tests to test the behaviour of new addTask method
with new function of checking duplicate tasks.

Some tests should be added to ensure that its behaviour
is consistent with what is expected. These tests can also be
used in future whenever there is a code change as a verification.

Let's add tests when duplicate todo, deadline
and event tasks are added.

carriezhengjr committed:
Resolve checkstyle errors

GitHub prompts that some checks were not successful,
which are caused by checkstyle errors.

The errors should be fixed for successful build and check
with Gradle on GitHub as well as successful build
when checkstyle command is keyed in manually into terminal.

Let's fix checkstyle errors
to check if Gradle can build successfully now.

carriezhengjr committed:
A-BetterGui: improve GUI

There is no prompt for the commands that user can enter.
The error messages also appear in the same style as other messages.

Adding buttons so that user guide is shown after clicking
could help to guide the user.
Highlighting errors with a different style
could catch user's attention.
These changes could improve UX.

Let's
* highlight error messages in a different format
* add user guide with quick start and detailed guide options

carriezhengjr committed:
A-Personality: add a unique personality

CaCa has a lovely personality but more can be done to show it.

Adding features like emojis can show CaCa's personality better
and make CaCa unique.

Let's add some emojis.

carriezhengjr committed:
A-MoreErrorHandling: add more error messages

User could mark or unmark a task repeatedly without any warning
for tasks that are already marked or unmarked.

Warning messages could be displayed to improve this function
to prompt user about it.

Let's add warning messages when user wants to mark or unmark
tasks that have already been marked or unmarked previously.

carriezhengjr committed:
A-MoreTesting: add more automated JUnit tests

There is no test to check the behaviours of the
newly modified mark and unmark task.

Adding some tests to check for its behaviour could ensure that
this function is working as intended.

Let's write tests for marking and unmarking a test.

carriezhengjr committed:
C-Help: add in-app guidance

There is in-app guidance of quick start or detailed guidance
when user presses the respective button, but there is no command
to the guidance if user prefers typing a command.

Other than clicking the buttons for guide,
a help command that leads to the guide could also help
users to navigate this program more easily.

Let's add a help command for users to access the detailed help page.

carriezhengjr committed:
C-Help: edit help response

Help response from CaCa using help command has a different
interface as the detailed guide access through the button.
They are not consistent.

This might confuse the user. Since both leads to the same help
page, we could make them to have the same interface.

Let's align the guide interfaces.

carriezhengjr committed:
Add error handling for empty task list

The error message shows incorrect index range suggestion
when task list is empty.

This is a bug and will mislead the user.

Let's fix this bug by adding a separate error message
when task list is empty.

carriezhengjr committed:
Improve CaCa's personality

CaCa's personality is not consistent throughout
as there is no emoji for some commands.

Adding more emojis and ensure consistency could improve UX.

Let's add more emojis that represents CaCa's personality.

carriezhengjr committed:
A-UserGuide: add a user guide

There is no guide for the usage of CaCa.

Adding a user guide will introduce the app and guide
users in setting up the app especially for the first time.

Let's add a user guide.

carriezhengjr committed:
Update image in README.md

The screenshot of CaCa application is too large 
when viewed on GitHub.

Resizing it with a specified width will 
make it more aesthetically pleasing.

Let's use another method to add image with width.

carriezhengjr committed:
Update README.md

Screenshot could not be seen on GitHub page.

Let's change the way screenshot is inserted.

carriezhengjr committed:
Update README.md

Update screenshot image link
to GitHub Page link.

carriezhengjr committed:
Update README.md

Minor changes to some words and user guide link.

carriezhengjr committed:
Update README.md and User Guide

Improve on the style of both READMEs to attract readers attention
and provide easier navigability.

Let's
* edit the fonts and some contents to improve on the main README.md
* add links to lead each feature to the respective section

carriezhengjr committed:
Update README and emoji

An emoji used to illustrate CaCa's personality could not be
displayed properly on other OS, i.e. Windows.
There is also a data folder created when users run CaCa for
the first time.

It should be changed to another emoji so that it can appear
normally as expected.
README could include more information about user's data
such as where it is saved and how user can access it.

Let's
* change the emoji to a working one
* update README.md to include more details

carriezhengjr committed:
Change emoji used

The original emojis have ASCII encoding issues.

This would affect the display of CaCa response and should be
fixed to ensure expected output as display.

Let's
* change all the emojis to simple ASCII
* update tests outputs
* update README expected outputs

carriezhengjr committed:
Update Ui screenshot

Update Ui screenshot to show the new emojis.

carriezhengjr committed:
Convert images to png, disable resizing

Images could not be loaded on Linux,
possibly because they are in jpg format.
Window could be resized, but the proportion of components is incorrect.

Converting image format from jpg to png could possibly solve the issue.
Disable resizing of window provides a quick fix
of window resizing problem.

Let's
* convert image format to png
* disable window resizing

carriezhengjr committed:
Remove emoji for title

Emoji of robot could not be shown on some OS.

It would be better to exclude it so that weird symbols will
not be shown.

Let's remove the emoji.
